contracts/RockPaperScissors.sol

```diff
@@ -0,0 +1,268 @@
+//SPDX-License-Identifier: MIT
+pragma solidity >= 0.6.0 < 0.7.0;
+
+import "./access/Ownable.sol";
+import "./utils/Pausable.sol";
+
+import {SafeMath} from "./math/SafeMath.sol";
+
+contract RockPaperScissors is Ownable, Pausable {
+
+    using SafeMath for uint;
+
+    mapping(address => uint) public balances;
+    mapping(bytes32 => Game) public games;
+
+    address private constant NULL_ADDRESS = address(0);
+    string private constant ONLY_CREATOR_CALL_MSG = "Only creator can call";
+    string private constant GAME_NOT_FOUND_MSG = "Game not found";
+    string private constant INVALID_MOVE_MSG = "Invalid move";
+
+    uint256 constant FORFEIT_WINDOW = 12 hours;
+    uint256 constant FEE_PERCENTAGE = 10;
+
+    struct Game {
+        address creator;
+        uint256 stake;
+        bytes32 moveHash;
+        address opponent;
+        uint8 opponentMove;
+        uint256 expiryDate;
+    }
+
+    event Deposit(
+        address player,
+        uint256 amount
+    );
+
+    event GameCreated(
+        address indexed creator,
+        bytes32 indexed gameToken,
+        uint256 stake
+    );
+
+    event OpponentPlays(
+        address indexed opponent,
+        bytes32 indexed token,
+        uint8 move
+    );
+
+    event Draw(
+        bytes32 indexed token
+    );
+
+    event WinnerRewarded(
+        bytes32 indexed gameToken,
+        address indexed winner,
+        address indexed loser,
+        uint256 winnings,
+        uint8 creatorMove,
+        uint8 opponentMove
+    );
+
+    event WithDraw(
+        address indexed withdrawer,
+        uint amount
+    );
+
+    event  CreatorEndsGame(
+        address indexed player,
+        bytes32 indexed gameToken,
+        uint256 refund
+    );
+
+    event  ForfeitPaid(
+        address indexed player,
+        bytes32 indexed gameToken,
+        uint256 forfeit
+    );
+
+    function deposit() external payable whenNotPaused {
+
+        balances[msg.sender] = balances[msg.sender].add(msg.value);
+
+        emit Deposit(msg.sender, msg.value);
+    }
+
+    function create(bytes32 playerMoveHash, uint stake) external payable whenNotPaused  returns(bytes32){
+
+        require(playerMoveHash != bytes32(0), "Player move cannot be empty");
+        require(stake > 0, "Stake cannot be zero");
+
+        bytes32 gameToken = keccak256(abi.encodePacked(msg.sender, block.timestamp));
+
+        Game storage game = games[gameToken];
+        require(game.creator == NULL_ADDRESS, "Token already in use");
```

Comment:
This prevents 2 games per block per creator. It's a constraint, although not a big one. Was this the intent?

Reply:
Also, do we agree that you currently allow password reuse?

Reply:
> This prevents 2 games per block per creator. It's a constraint, although not a big one. Was this the intent?

Sorry, I don't understand what that means. I understood that when create is called, it creates a new block so cannot be called twice per block.

My intention is that struct record can never be reused. Very unlikely given the that the gameToken will never be the same. Therefore, I think the require() is redundant.

Reply:
> Also, do we agree that you currently allow password reuse?

Yes. I was considering if there is a way to look at the historical games and for the opponent to reverse-engineer the creator's moveHash. The secret/password provides some entropy and the non-deterministic component thereby acting like a salt.

There is nothing to prevent reuse of 'secret' but hashing with 'gameToken' would ensure that each hash is always unique. I realise adding 'msg.sender' is probably not providing any entropy value as it will the same every time so could be removed.

I could perhaps have an additional mapping in storage to store past secrets and prevent their reuse but does reuse of the secret expose the creator's moves? I considered very hard but cannot see how.

Reply:
>  I understood that when create is called, it creates a new block

That's only true of Ganache. In a real setup, the transactions are arranged and many transactions go into a single block.

> My intention is that struct record can never be reused.

That's what the `require` is for. That is independent from how you create the `gameToken`.

> Very unlikely given the that the gameToken will never be the same

It is very likely in a non-Ganache setup. The player just sends 2 `create` transactions that get mined in the same block. The second transaction will revert. Not a good look.

> Therefore, I think the require() is redundant.

Not redundant because in the point above, the second transaction would not revert but instead overwrite the first, thereby burning other people's money.

> The reason I don't want to reuse the gameToken in subsequent games is because it guarantees uniqueness in the creator's moveHash.

No it does not guarantee. It only guarantees the uniqueness of the `gameToken`, which is not really useful.

> The secret/password provides some entropy and the non-deterministic component thereby acting like a salt.

I don't understand what salt you are talking about. If the password and salt were used once, that's it, they are known.

> There is nothing to prevent reuse of 'secret' but hashing with 'gameToken' would ensure that each hash is always unique

That `gameToken` happens to be unique is not worthwhile. An attacker needs to find out what's the move behind a `moveHash`. Anything that is happening to `moveHash` after that is not part of what needs to be known.

> I realise adding 'msg.sender' is probably not providing any entropy value as it will the same every time so could be removed.

If you remove it then the same password used by Alice and Bob will result in the same `moveHash`, a very poor safety.

> I could perhaps have an additional mapping in storage to store past secrets and prevent their reuse

You would store the secrets? But that's possible only after disclosure. If it so happens that alice started a game with bob and started another one with carol, and none has been revealed, and she used the same password, then the first time she reveals, the other player already knows what she player and so can win 100%.

> does reuse of the secret expose the creator's moves?

Absolutely. Alice creates with `hash(Rock, "password")`, then reveals `Rock, "password"`. Later on, she creates another game with `hash(Rock, "password")`. Well, that's it, everybody knows that it is `Rock`.

Reply:
Thank you for the explanation and your patience.

I understand the block issue better now.

> No it does not guarantee. It only guarantees the uniqueness of the `gameToken`, which is not really useful.

But if the game token is unique and it is used in the keccak256 in hashPlayerMove() then I thought the hash generated should also be unique so when paired with the creator's move, there will never be a match to compare and determine the corresponding move.

> > does reuse of the secret expose the creator's moves?
> 
> Absolutely. Alice creates with `hash(Rock, "password")`, then reveals `Rock, "password"`. Later on, she creates another game with `hash(Rock, "password")`. Well, that's it, everybody knows that it is `Rock`.

What I am struggling with is that if Alice creates another game with `hash(Rock, "password"), it will also have the unique game token incorporated into the hash, making the hash unique so it cannot be matched with the previous 'Rock' move.

I think this is an area I am really struggling to understand fundamentally. I accept that this is insecure albeit I need to go away and understand better the theory that explains why. In the meantime I will try to think of a smarter way to improve the security.

Reply:
> But if the game token is unique and it is used in the keccak256 in hashPlayerMove()

Oof I had missed that bit. That is bad. And untested obviously. What you are doing is you are forcing the game creator to have their transaction mined in a future block whose timestamp they have to guess beforehand. You can easily imagine that the reveal will "never" work, unless the game creator is a miner itself. Draw a timeline and put along it the times at which the game creator calls the functions up to `create`.

Reply:
Very embarrassing. I changed the design and completely overlooked this. 🤦‍♂️

Reply:
I have an idea to prevent reuse of 'secret'. The 'secret' is used to create the 'games' mapping key and attempts to use the same key again will cause create() to revert.

I will rename 'gameToken' to 'gameHashKey'. I think using the word 'token' is not ideal in this context.

Reply:
Let's see